src/hevm/hevm-cli/hevm-cli.hs

```diff
@@ -355,15 +372,32 @@ findJsonFile Nothing = do
         , intercalate ", " xs
         ]
 
-dappTest :: UnitTestOptions -> Mode -> String -> IO ()
-dappTest opts _ solcFile =
+concatMapM :: Monad m => (a -> m [b]) -> [a] -> m [b]
+concatMapM op = foldr f (pure [])
+    where f x xs = do x <- op x; if null x then xs else do xs <- xs; pure $ x++xs
+
+dappTest :: UnitTestOptions -> Mode -> String -> Maybe String -> IO ()
+dappTest opts _ solcFile cache =
   readSolc solcFile >>=
     \case
-      Just (contractMap, cache) -> do
+      Just (contractMap, sourceCache) -> do
         let matcher = regexMatches (EVM.UnitTest.match opts)
             unitTests = (findUnitTests matcher) (Map.elems contractMap)
-        results <- mapM (runUnitTestContract opts contractMap cache) unitTests
-        when (any (== False) results) exitFailure
+
+        results <- concatMapM (runUnitTestContract opts contractMap sourceCache) unitTests
+        let (passing, vms) = unzip results
+
+        case cache of
+          Nothing ->
+            pure ()
+          Just path ->
+            -- merge all of the post-vm caches and save into the state
```

Comment:
If the cache was shared between unitTests we'd have better execution for the initial run and runs withoug `--cache`

Reply:
Yes, agree, was just a bit more work to implement.